### PR TITLE
docs: add neptune docs

### PR DESCRIPTION
**Pull request by friendtocephalopods**
### Summary of Changes

Add docs, add sample terraform file.

All credit to @dsimms for the documentation and terraform file. Thanks Dan!!

### CheckList

Make sure you have checked **all** steps below to ensure a timely review.

- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.

### CODE REVIEW

File: amundsen_gremlin/neptune_bulk_loader/api.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 )
 from gremlin_python.process.anonymous_traversal import traversal
 from gremlin_python.process.graph_traversal import GraphTraversalSource
+from neptune_python_utils.endpoints import Endpoints, RequestParameters
```
Comment (friendtocephalopods) [inline amundsen_gremlin/neptune_bulk_loader/api.py:27]: Due to the wacky way we add neptune_python_utils as a submodule here, if you forget to `setup.py install` as I apparently did last time, isort will think it's part of this package. It belongs up here.

File: docs/docs/neptune_for_amundsen.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+into S3 and then invoking the Neptune Bulk Loader API to load those and check on status.  In order
+for the Neptune cluster to be able to access the files, there must be a S3 VPC endpoint in the VPC
+in which the Neptune cluster runs.  (It feels like there's a long story behind this.)  Further,
+the Neptune cluster needs an assumable that allows it to access the files upon which the Bulk
+Loader is invoked.  (This is potentially very simple... but also potentially extremely complicated
+because the policies of the S3 VPC Endpoint, the Bucket permissions on the S3 Bucket used, the
+per-Object permissions on the files in the Bucket, as well as those of the assumed Role.)
+
+The terraform example creates: the S3 VPC Endpoint, a role for Neptune which gives it readonly
+access to all S3 buckets and object (in that account), and a bucket.
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: tiny nit: object -> object**s**.

```suggestion
access to all S3 buckets and objects (in that account), and a bucket.
```

File: docs/docs/neptune_for_amundsen.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+### Create Neptune cluster and instances
+
+Possibly, you will create a Neptune Subnet Group.  If you're using the default VPC, a Neptune
+Subnet Group already exists.  Otherwise, you create one. 
+
+Then Create a cluster and (at least) one instance, using the Security Group and assumable Role.
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: tiny nit: Create -> create

```suggestion
Then create a cluster and (at least) one instance, using the Security Group and assumable Role.
```

File: docs/docs/neptune_for_amundsen.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+The cluster's `cluster_identifier` and the instance's `identifier` will become part of the DNS
+names so choose them appropriately.  The `instance_class` configures the amount of RAM, CPU, and
+IO available to the instance, as well is being the primary factor in cost.  (Chosing this is
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: "as well is being" -> "as well as being"

```suggestion
IO available to the instance, as well as being the primary factor in cost.  (Chosing this is
```

File: docs/docs/neptune_for_amundsen.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+experimental, and beyond the scope of this guide.)  
+
+There are other interesting paramters related to admininstration: the maintenance window,
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: "paramters" -> "parameters"
typo: "admininstration" -> "administration"
```suggestion
There are other interesting parameters related to administration: the maintenance window,
```

File: docs/docs/neptune_for_amundsen.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+automatic minor version upgrades, deletion protection, backup retentoin, encryption-at-rest and
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: "retentoin" -> "retention"
```suggestion
automatic minor version upgrades, deletion protection, backup retention, encryption-at-rest and
```

File: docs/docs/neptune_for_amundsen.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+
+## Neptune setup 
+
+### Summary
+
+* Create a Security Group that allows connections to your Neptune instance.  (safely)
+* Create a VPC endpoint for S3, a bucket, and a Role by which Neptune can access that.
+* Create Neptune cluster and instances.
+* Create user(s) and give it(them) access to Neptune and S3.
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: style nit: I find the "it(them)" stuff confusing. Can you pick one? 
```suggestion
* Create user(s) and give them access to Neptune and S3.
```

File: docs/docs/neptune_for_amundsen.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+There are other interesting paramters related to admininstration: the maintenance window,
+automatic minor version upgrades, deletion protection, backup retentoin, encryption-at-rest and
+key management for that, default query execution timeouts.  (These are also beyond the scope of
+this.)
+
+
+The terraform example configures a Neptune Subnet Group, a Neptune Cluster, and 2 Neptune Instances.
+
+
+### Create user(s) and give it(them) access to Neptune and S3
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]:
```suggestion
### Create user(s) and give them access to Neptune and S3
```

File: docs/docs/neptune_for_amundsen.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+The metadata and dataloader processes require access to Neptune and S3.  In the most direct setup,
+each would use a distinct IAM user and receive just those permissions required to access Neptune
+and S3.  An another, you might run the metadata service and dataloader processes in EC2 and you
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: "An another" -> "In another"
```suggestion
and S3.  In another, you might run the metadata service and dataloader processes in EC2 and you
```

File: docs/docs/neptune_for_amundsen.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+would grant whatever Role they assume the required permissions for Neptune and S3.
+
+The terraform example creates a Policy which gives the metadata and dataloader services sufficient
+access, and creates 2 Users and gives attaches the Policy.
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: "gives attaches" -> "attaches"
```suggestion
access, and creates 2 Users and attaches the Policy.
```

File: docs/docs/neptune_for_amundsen.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+### Create public endpoint for Neptune
+
+Neptune does not have public endpoints.  The state of the art seems to be using
+a [loadbalancer](
+https://github.com/aws-samples/aws-dbs-refarch-graph/tree/master/src/connecting-using-a-load-balancer)
+as well as glue to allow the mismatch in hostnames.
+
+
+The terraform sets up a publicly available NLB -- albeit statically configured, so is only suitable
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: small nit
```suggestion
The terraform sets up a publicly available NLB -- albeit statically configured, so it is only suitable
```

File: docs/docs/neptune_for_amundsen.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+for development or other deployments where one could tolerate downtime as instances are
+reconfigured.  (Or you could apply the CloudWatch + Lambda approach and minimize that downtime.)
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: I don't think the "CloudWatch + Lambda" approach is described elsewhere in this doc, seems out of scope and maybe not worth mentioning here. WDYT?

```suggestion
reconfigured. 
```

File: docs/docs/neptune_for_amundsen.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+
+### Configure amundsen-gremlin
+
+amundsen-gremlin needs to know about some of the resources you configured:
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]:
```suggestion
amundsen-gremlin needs to know about some of the resources you configured with Terraform:
```

File: docs/docs/neptune_for_amundsen.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+  ```
+
+
+## Other Neptunic Considerations
+
+
+## IAM authentication
+
+You should use IAM authentication.  However, it comes with some mild complications.  The gremlin
+transport is usually websockets, and the requests-aws4auth library we use elsewhere is for requests,
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: "requests" as a term here is pretty overloaded. I think hyperlinking to the requests _library_ would help disambiguate. WDYT? 

```suggestion
transport is usually websockets, and the requests-aws4auth library we use elsewhere is for [requests](https://requests.readthedocs.io/en/master/),
```

File: docs/docs/neptune_for_amundsen.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+which does not support websockets at all.  In the [amazon-neptune-tools](
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: I think either "the" from "In the amazon-neptune-tools, there is..." sentence should be dropped, or the word "repo" added at the end (e.g. "in the amazon-neptune-tools repo, there is..."

```suggestion
which does not support websockets at all.  In [amazon-neptune-tools](
```

File: docs/docs/neptune_for_amundsen.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+https://github.com/awslabs/amazon-neptune-tools/tree/master/neptune-python-utils), there is an
+example for authenticating gremlin websocket connection which is reused herein.  (But is technically
+an example so may or may not work in your circumstances.)
+
+
+## How to get a gremlin console for Neptune
+
+They (AWS) [have a goodrecipe](
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: "goodrecipe" -> "good recipe". Also rewording for simplicity.
```suggestion
AWS [has a good recipe](
```

File: docs/docs/neptune_for_amundsen.md [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+https://docs.aws.amazon.com/neptune/latest/userguide/iam-auth-connecting-gremlin-java.html).
+
+You may also be interested in using[Jupyter/Sagemaker](
```
Comment (cpu) [inline docs/docs/neptune_for_amundsen.md]: typo: space missing between "using" and "Jupyter/Sagemaker"

```suggestion
You may also be interested in using [Jupyter/Sagemaker](
```

File: docs/docs/terraform/main.tf [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+  name       = "service"
+  subnet_ids = data.aws_subnet_ids.neptune_contained_herein.ids
+}
+
+resource "aws_neptune_cluster" "service" {
+  count = 1
+  engine = "neptune"
+  cluster_identifier = var.service
+  iam_database_authentication_enabled = true
+  # engine_version default
```
Comment (cpu) [inline docs/docs/terraform/main.tf]: Is this configuration syntax that's been commented out, or a comment explaining something? If the former maybe delete it outright. If the later, needs more context.

File: docs/docs/terraform/main.tf [hunk 19]
```diff
@@ -0,0 +1,466 @@
...
+  identifier = aws_neptune_cluster.service[count.index].cluster_identifier
+  instance_class = "db.t3.medium"
+  neptune_parameter_group_name = aws_neptune_parameter_group.service.name
+  auto_minor_version_upgrade = true
+}
+
+# Supposing you need a public Neptune endpoint, that doesn't happen directly out of the box.
+# Our solution is this:
+#   * put an NLB (network load balancer) in front of each cluster
+#   * the points at an IP address and TCP port (the Neptune server)
```
Comment (cpu) [inline docs/docs/terraform/main.tf]: typo: "the points" -> "that points"
```suggestion
#   * that points at an IP address and TCP port (the Neptune server)
```

File: docs/docs/terraform/main.tf [hunk 20]
```diff
@@ -0,0 +1,466 @@
...
+#   * the Neptune server terminates TLS
+#   * but because the address we connect to is different that what it advertises, we install a special
```
Comment (cpu) [inline docs/docs/terraform/main.tf]: nit: I wouldn't say "advertises" here since that makes me think about DNS. Instead I think it might be clearer to talk about the Neptune server's end entity certificate not containing a subject that matches the address being connected to.

```suggestion
#   * but because the address we connect to is not present as a subject in the Neptune TLS certificate, we install a special
```

File: docs/docs/terraform/main.tf [hunk 21]
```diff
@@ -0,0 +1,466 @@
...
+#     SSLContext (for Gremlin/GLV) or HostAdapter (for requests)
+# You have questions:
```
Comment (cpu) [inline docs/docs/terraform/main.tf]: Really appreciate these "alternatives considered"/FAQ items.

File: docs/docs/terraform/main.tf [hunk 22]
```diff
@@ -0,0 +1,466 @@
...
+}
+
+resource "aws_iam_user" "dataloader" {
+  name = "dataloader"
+}
+
+locals {
+  # put the GPG public key you wish to encrypt the secrets with here
+  # gpg1 --no-default-keyring --keyring ./whatever-key-ring --export whatever-key-id  | base64
+  pgp_key = "...."
```
Comment (cpu) [inline docs/docs/terraform/main.tf]: Should this be called out in the `neptune_for_amundsen.md` doc someplace?

**Review by cpu — CHANGES_REQUESTED** [review]
Mostly editorial changes & typos to flag, otherwise LGTM. Thanks @friendtocephalopods !

File: docs/docs/neptune_for_amundsen.md (hunk 12 above)
Comment (friendtocephalopods) [reply docs/docs/neptune_for_amundsen.md]: I think it's possibly useful as a breadcrumb or exercise for the reader if one cannot tolerate such downtime.

File: docs/docs/terraform/main.tf (hunk 18 above)
Comment (friendtocephalopods) [reply docs/docs/terraform/main.tf]: Looks like [syntax](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/neptune_cluster#engine_version) commented out, IIUC you can pick which release of neptune you want. It's not super clear to me why you'd want a specific engine version, so I'll probably remove.

File: docs/docs/neptune_for_amundsen.md (hunk 12 above)
Comment (cpu) [reply docs/docs/neptune_for_amundsen.md]: Sounds good to me. I noticed after the fact that some comments in the terraform code do dig into this a little bit more so it seems reasonable to leave 👍

File: docs/docs/terraform/main.tf (hunk 22 above)
Comment (friendtocephalopods) [reply docs/docs/terraform/main.tf]: Hmmm, maybe? It feels like the neptune for amundsen doc is more general guidance, whereas this terraform is more of an example you can use or model your impl after (you don't need to use a pgp_key here if you would rather not)

**Review by cpu — APPROVED** [review]

File: docs/docs/terraform/main.tf (hunk 22 above)
Comment (cpu) [reply docs/docs/terraform/main.tf]: Seems reasonable to leave as-is 👍

**Comment by Samlinxia** [comment]
@friendtocephalopods , @cpu  thanks for working on this.  We are currently investigating to migrate our backend from Neo4j to AWS Neptune for Amundsen.  Is the AWS Neptune ready to use in production since it seems didn't announce in publicly release notes?  Thanks!

**Comment by cpu** [comment]
:wave: Hi @Samlinxia I think there is still work required to be done to get the databuilder to support Neptune. See [this RFC](https://github.com/amundsen-io/rfcs/pull/13). In the #gremlin channel of the Amundsen slack the RFC author mentioned they hoped to have a PR open in early Dec but I can't speak personally to where that's at. Hope that helps!
